crates/adapters/src/integrated/delta_table/input.rs

```diff
@@ -210,9 +217,186 @@ impl DeltaTableInputEndpointInner {
+
+        let timestamp_field = schema.field(timestamp_column).unwrap();
+
+        // The following unwraps are safe, as validated in `validate_timestamp_column`.
+        let lateness = timestamp_field.lateness.as_ref().unwrap();
+
+        // Query the table for min and max values of the timestamp column that satisfy the filter.
+        let bounds_query =
+            format!("select * from (select cast(min({timestamp_column}) as string) as start_ts, cast(max({timestamp_column}) as string) as end_ts from snapshot {}) where start_ts is not null",
+            if let Some(filter) = &self.config.snapshot_filter {
+                format!("where {filter}")
+            } else {
+                String::new()
+            });
+
+        let bounds = self.execute_query_collect(&bounds_query).await?;
+
+        if bounds.len() != 1 || bounds[0].num_rows() != 1 {
+            info!(
+                "delta_table {}: initial snapshot is empty; the Delta table contains no records{}",
+                &self.endpoint_name,
+                if let Some(filter) = &self.config.snapshot_filter {
+                    format!(" that satisfy the filter condition '{filter}'")
+                } else {
+                    String::new()
+                }
+            );
+            return Ok(());
+        }
+
+        if bounds[0].num_columns() != 2 {
+            // Should never happen.
+            return Err(anyhow!(
+                    "internal error: query '{bounds_query}' returned a result with {} columns; expected 2 columns",
+                    bounds[0].num_columns()
+                ));
+        }
+
+        let min = bounds[0]
+            .column(0)
+            .as_string_opt::<i32>()
+            .ok_or_else(|| anyhow!("internal error: cannot retrieve the output of query '{bounds_query}' as a string"))?
+            .value(0)
+            .to_string();
+
+        let max = bounds[0].column(1).as_string::<i32>().value(0).to_string();
+
+        info!(
+            "delta_table {}: reading table snapshot in the range '{min} <= {timestamp_column} <= {max}'",
+            &self.endpoint_name,
+        );
+
+        let min = Self::timestamp_to_sql_expression(&timestamp_field.columntype, &min);
+        let max = Self::timestamp_to_sql_expression(&timestamp_field.columntype, &max);
+
+        let mut start = min.clone();
+        let mut done = "false".to_string();
+
+        while &done != "true" {
+            // Evaluate SQL expression for the new end of the interval.
+            let end = self
+                .execute_singleton_query(&format!("select cast(({start} + {lateness}) as string)"))
+                .await?;
+            let end = Self::timestamp_to_sql_expression(&timestamp_field.columntype, &end);
+
+            // Query the table for the range.
+            let mut range_query =
+                format!("select * from snapshot where {timestamp_column} >= {start} and {timestamp_column} < {end}");
```

Comment:
I see why it can't be zero
where is that validated?

Reply:
there's a validation function that checks that.